[Kobot7] iP

Kobot7 opened the pull request.

Kobot7 committed:
Rename Duke to Kobot. Greet user and exit.

Kobot7 committed:
Add echo feature

Kobot7 committed:
Add task list feature

Kobot7 committed:
Add feature to mark/unmark tasks

Kobot7 committed:
Add Todos, Events, Deadlines

Kobot7 committed:
Add Automated Text UI Testing

e0958902 commented on src/main/java/Ui.java:
```diff
@@ -0,0 +1,24 @@
+            + "##    | |\\  \\ (_) | |_) | (_) | |_     ##\n"
+            + "##    \\_| \\_/\\___/|_.__/ \\___/ \\__|    ##\n"
+            + "##                                     ##\n"
+            + "#########################################\n";
```
Good use of line separation to ensure readability.

e0958902 commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,107 @@
+        default:
+            System.out.println("Invalid command. Please try again.");
+            break;
+        }
```
Good use of switch statement with proper indentations.

e0958902 commented on src/main/java/Command.java:
```diff
@@ -0,0 +1,107 @@
+     * @param input User input.
+     */
+    public void parseCommand(String input) {
+        String[] parts = input.split(" ", 2);
```
Perhaps you could change "2" into a variable name that is relevant such as "twoSubString"

e0958902 commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,82 @@
+            return;
+        }
+        taskList[index].markAsNotDone();
+        System.out.println("Okay, I've marked this task as not done yet: " + taskList[index].getDescription());
```
Perhaps you could wrap to another line although the line length does not exceed the limit.

e0958902 reviewed (commented):
Additional comments to add on.

Kobot7 committed:
Add error handling for empty and missing arguments

Kobot7 committed:
Add error handling for marking/unmarking invalid tasks

Kobot7 committed:
Fix printing of error message when marking/unmarking invalid tasks

Kobot7 committed:
Add error handling for unexpected signals

Kobot7 committed:
Merge branch 'branch-Level-5'

Kobot7 committed:
Divide classes into packages

Kobot7 committed:
Merge branch 'branch-A-Packages'

json-bot-boop commented on src/main/java/kobot/Command.java:
```diff
@@ -0,0 +1,164 @@
+        String[] parts = input.split(" ", 2);
+        this.command = parts[0].toUpperCase();
+        if (parts.length > 1) {
+            this.arguments = parts[1];
```
Try to avoid magic number

json-bot-boop commented on src/main/java/kobot/Command.java:
```diff
@@ -0,0 +1,164 @@
+            break;
+
+        default:
+            System.out.println("Invalid command. Please try again.");
```
Try to avoid magic string, maybe can pre-define them as final variables

json-bot-boop commented on src/main/java/kobot/task/TaskList.java:
```diff
@@ -0,0 +1,90 @@
+     *
+     * @param description Description of the task.
+     */
+    public void addToDo(String description) {
+        taskList[taskCount] = new ToDo(description);
+        System.out.println("Task has been added to list:");
+        System.out.println(taskList[taskCount]);
+        taskCount++;
+    }
+
+    /**
+     * Adds a task that needs to be done by a specific date/time to the task list.
+     *
+     * @param description Description of the task.
+     * @param by Date/Time that the task has to be completed by.
+     */
+    public void addDeadline(String description, String by) {
+        taskList[taskCount] = new Deadline(description, by);
+        System.out.println("Deadline has been added to list:");
+        System.out.println(taskList[taskCount]);
+        taskCount++;
+    }
+
+    /**
+     * Adds a task that starts and ends at a specific date/time.
+     *
+     * @param description Description of the task.
+     * @param from Date/Time that the task starts.
+     * @param to Date/Time that the task ends.
+     */
+    public void addEvent(String description, String from, String to) {
+        taskList[taskCount] = new Event(description, from, to);
+        System.out.println("Event has been added to list:");
+        System.out.println(taskList[taskCount]);
+        taskCount++;
+    }
```
Maybe can extract the last 3 lines out from these 3 methods and put into a new method to make it less repetitive

json-bot-boop commented on src/main/java/kobot/Command.java:
```diff
@@ -0,0 +1,164 @@
+        this.isExit = true;
+    }
+
+    public Boolean getIsExit() {
```
Boolean method should be named to sound like boolean

json-bot-boop commented on src/main/java/kobot/Kobot.java:
```diff
@@ -0,0 +1,31 @@
+    public Kobot() {
+        this.taskList = new TaskList();
+        this.ui = new Ui();
+        this.in = new Scanner(System.in);
```
Maybe can have a better variable name like "scanner" instead of "in"

json-bot-boop commented on src/main/java/kobot/task/TaskList.java:
```diff
@@ -0,0 +1,90 @@
+     *
+     * @param index Index of the task to mark as done.
+     */
+    public void markTask(int index) throws NumberFormatException {
+        if (index >= taskCount || index < 0) {
+            System.out.println("Failed to mark item " + (index + 1) + ". Index out of range.");
+            return;
+        }
+        taskList[index].markAsDone();
+        System.out.println("Nice! I've marked this task as done: " + taskList[index].getDescription());
+    }
+
+    /**
+     * Marks a specified task as not done.
+     *
+     * @param index Index of the task to mark as not done.
+     */
+    public void unmarkTask(int index) throws NumberFormatException {
+        if (index >= taskCount || index < 0) {
+            System.out.println("Failed to unmark item " + (index + 1) + ". Index out of range.");
+            return;
+        }
+        taskList[index].markAsNotDone();
+        System.out.println("Okay, I've marked this task as not done yet: " + taskList[index].getDescription());
+    }
```
Maybe can combine these 2 methods together and use a new method to be able to toggle the boolean variable in the object instead of setting it manually.

json-bot-boop commented on src/main/java/kobot/Ui.java:
```diff
@@ -0,0 +1,79 @@
+        System.out.println("Bye, hope to see you again!");
+    }
+
+    public String receiveInput(Scanner in) throws NoSuchElementException {
+        printLineDivider();
+        System.out.print("> ");
+        String input = "";
+        
+        try {
+            input = in.nextLine();
+        } catch (NoSuchElementException exception) {
+            System.exit(0);
+        }
+        
+        printLineDivider();
+        return input;
+    }
```
Is there a exception that will be thrown out?

json-bot-boop commented on src/main/java/kobot/Command.java:
```diff
@@ -0,0 +1,164 @@
+            } catch (IllegalArgumentException exception) {
+                ui.printEmptyArgumentErrorMessage();
+                ui.printDeadlineCommandUsage();
+            }
+            break;
+
+        case "EVENT":
+            try {
+                parseEventArguments();
+                taskList.addEvent(this.description, this.from, this.to);
+            } catch (NullPointerException| IndexOutOfBoundsException exception) {
+                ui.printMissingArgumentErrorMessage();
+                ui.printEventCommandUsage();
+            } catch (IllegalArgumentException exception) {
+                ui.printEmptyArgumentErrorMessage();
+                ui.printEventCommandUsage();
+            }
+            break;
+
+        case "MARK":
+            try {
+                taskList.markTask(Integer.parseInt(this.arguments) - 1);
+            } catch (NumberFormatException exception ) {
+                ui.printMarkCommandUsage();
+            }
+            break;
+
+        case "UNMARK":
+            try {
+                taskList.unmarkTask(Integer.parseInt(this.arguments) - 1);
+            } catch (NumberFormatException exception ) {
+                ui.printUnmarkCommandUsage();
+            }
+            break;
+
+        default:
+            System.out.println("Invalid command. Please try again.");
+            break;
+        }
+    }
```
Try to avoid long methods, abstract some methods out if possible

json-bot-boop reviewed (commented):
Try to abstract some methods out but overall looks okay except some code quality and standard issue.

Kobot7 committed:
Add feature to delete tasks

Kobot7 committed:
Save tasks in local storage

Kobot7 committed:
Merge branch 'branch-Level-6'

Kobot7 committed:
Merge branch 'branch-Level-7'

# Conflicts:
#	src/main/java/kobot/Command.java
#	src/main/java/kobot/Ui.java
#	src/main/java/kobot/task/TaskList.java

Kobot7 committed:
Change Ui to static

Kobot7 committed:
Update code to conform to coding standard

- Replaced magic numbers and strings with constants
- Abstract long code to new methods
- Refactored command parsing and execution to individual classes

Kobot7 committed:
Add find task feature

Kobot7 committed:
Add Java-doc comments

Kobot7 committed:
Merge pull request #1 from Kobot7/branch-Level-9

Add find task feature

Kobot7 committed:
Merge pull request #2 from Kobot7/branch-A-JavaDoc

Add Java-doc comments

Kobot7 committed:
Update User Guide

Kobot7 committed:
Add more details to User Guide

Kobot7 committed:
Update JavaDoc
